### PR TITLE
Implement templating system

### PR DESCRIPTION
**Pull request by Kegsay**
This PR adds in a ``templating`` folder which contains the templating system for creating the specification. Currently, it is setup to generate RST for the room events section of the specification. It will be expanded to eventually consume the entire spec in later PRs, but this one is just to get the bare bones implementation working.

The templating system uses Jinja2 for templating, and python for arbitrary data processing. How it works is explained in the module docs for ``build.py``.

Example output:
```
``m.room.power_levels``
  Summary:
    Defines the power levels (privileges) of users in the room.
  Type:
    State Event
  Description:
    This event specifies the minimum level a user must have in order
    to perform a certain action. It also specifies the levels
    of each user in the room. If a ``user_id`` is in the 
    ``users`` list, then that ``user_id`` has the associated 
    power level. Otherwise they have the default level 
    ``users_default``. If ``users_default`` is not supplied,
    it is assumed to be 0. The level required to send a
    certain event is governed by ``events``, ``state_default``
    and ``events_default``. If an event type is specified in 
    ``events``, then the user must have at least the level
    specified in order to send that event. If the event type is
    not supplied, it defaults to ``events_default`` for Message
    Events and ``state_default`` for State Events.
  Required Keys:
    ``["ban", "events", "events_default", "kick", "redact", "state_default", "users"]``

  JSON Format::

    {
        "events_default": "<number>", 
        "users": {
            "<string>": "<number>"
        }, 
        "state_default": "<number>", 
        "redact": "<number>", 
        "ban": "<number>", 
        "events": {
            "<string>": "<number>"
        }, 
        "kick": "<number>", 
        "users_default": "<number>"
    }

  Example::

    {
        "events_default": 0, 
        "users": {
            "@example:localhost": 100
        }, 
        "state_default": 50, 
        "redact": 50, 
        "ban": 50, 
        "events": {
            "m.room.name": 100, 
            "m.room.power_levels": 100
        }, 
        "kick": 50, 
        "users_default": 0
    }

``m.room.member``
  Summary:
    The current membership state of a user in the room.
  Type:
    State Event
  Description:
    Adjusts the membership state for a user in a room. It is preferable to use
    the membership APIs (``/rooms/<room id>/invite`` etc) when performing
    membership actions rather than adjusting the state directly as there are a
    restricted set of valid transformations. For example, user A cannot force
    user B to join a room, and trying to force this state change directly will
    fail.
  Required Keys:
    ``["membership"]``

  JSON Format::

    {
        "membership": "<enum> (invite|join|knock|leave|ban)", 
        "avatar_url": "<string>", 
        "displayname": "<string>"
    }

  Example::

    {
        "membership": "join", 
        "avatar_url": "mxc://localhost/SEsfnsuifSDFSSEF#auto", 
        "displayname": "Alice Margatroid"
    }
```

### CODE REVIEW

File: event-schemas/schema/v1/m.room.aliases [hunk 1]
```diff
@@ -1,6 +1,8 @@
 {
     "$schema": "http://json-schema.org/draft-04/schema#",
     "type": "object",
+    "title": "Informs the room about what room aliases it has been given.",
+    "description": "This event is sent by a homeserver directly to inform of changes\nto the list of aliases it knows about for that room. The\n``state_key`` for this event is set to the homeserver which\nowns the room alias. The entire set of known aliases for the\nroom is the union of all the ``m.room.aliases`` events, one\nfor each homeserver. Clients **should** check the validity\nof any room alias given in this list before presenting it to\nthe user as trusted fact. The lists given by this event\nshould be considered simply as advice on which aliases\nmight exist, for which the client can perform the lookup to\nconfirm whether it receives the correct room ID.",
```
Comment (ara4n) [inline event-schemas/schema/v1/m.room.aliases]: having \n's in the description feels ugly - surely it should be up to the templating script to wrap the line sensibly?

**Comment by Kegsay** [comment]
Agreed.

**Comment by Kegsay** [comment]
Done.

File: templating/build.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
++-------+ |-+ === used to create ==> +----------- | === used to create ==> SPEC
+  +-------+ |                          +----------+
+   +--------+
+RAW DATA (e.g. json)                  Blobs of RST
+
+Units
+=====
+Units are random bits of unprocessed data, e.g. schema JSON files. Anything can
+be done to them, from processing it with Jinja to arbitrary python processing.
+They are dicts.
```
Comment (ara4n) [inline templating/build.py]: Presumably Units can also be fragments of RST? If so, we should spell this out.

**Comment by ara4n** [comment]
This LGTM, other than concern over the actual output template format.  In the example in comment 1, i'd expect to see per-field descriptions on the various keys, plus I'm wondering how more deeply nested schemas end up being displayed to the user.  Finally, I wonder if the <li/> style RST format for the template is actually the most legible - it always feels to me as if the list of event descriptions start to run together.  Either putting a header or subsection to separate them might help this, or possibly laying out the various fields in a table to make it more of an obvious reference rather than relatively unstructured bulletpoints?

File: templating/build.py (hunk 2 above)
Comment (Kegsay) [inline templating/build.py]: It's technically anything (just a `dict` of whose values could be RST). For clarity / preventing confusion, I decide to keep the distinction as "bits of data" vs "bits of RST" hence not spelling it out on purpose.

**Comment by Kegsay** [comment]
Will be adding in per field descriptions and shuffling the actual template in another PR. This current PR just has feature parity with the existing spec.

**Pull request merged**